🐛 BUG: Crash with full-width characters

### Describe the Bug

When there are full-width characters in the HTML code content, it causes the Astro language server to crash.

<img width="428" alt="Snipaste_2024-01-06_16-17-44" src="https://github.com/withastro/language-tools/assets/95267940/18de9e5c-6f9c-41b8-8035-c709750b3235">
<img width="447" alt="Snipaste_2024-01-06_16-28-25" src="https://github.com/withastro/language-tools/assets/95267940/e78a8c6b-d10c-4b40-907e-d6cbbb3e42b9">
<img width="394" alt="Snipaste_2024-01-06_16-30-23" src="https://github.com/withastro/language-tools/assets/95267940/35034320-bddb-457e-9c33-3dad691f7081">

```bash
Error: TypeError: Cannot read properties of undefined (reading 'ast').
panic: runtime error: index out of range [16] with length 16

goroutine 6 [running]:
github.com/withastro/compiler/internal/sourcemap.(*ChunkBuilder).GetLineAndColumnForLocation(0x4cbb88, {0x67})
	github.com/withastro/compiler/internal/sourcemap/sourcemap.go:632 +0x11
github.com/withastro/compiler/internal/printer.locToPoint(0x4cbab8, {0x67})
	github.com/withastro/compiler/internal/printer/print-to-json.go:123 +0x4
github.com/withastro/compiler/internal/printer.positionAt(0x4cbab8, 0x4a5540, {{0x39f01, 0x7}, 0x1})
	github.com/withastro/compiler/internal/printer/print-to-json.go:167 +0xd
github.com/withastro/compiler/internal/printer.renderNode(0x4cbab8, 0x4cb278, 0x4a5540, {{0x39f01, 0x7}, 0x1})
	github.com/withastro/compiler/internal/printer/print-to-json.go:209 +0x9
github.com/withastro/compiler/internal/printer.renderNode(0x4cbab8, 0x4cb630, 0x4a5400, {{0x39f01, 0x7}, 0x1})
	github.com/withastro/compiler/internal/printer/print-to-json.go:267 +0x4b
github.com/withastro/compiler/internal/printer.renderNode(0x4cbab8, 0x4cb928, 0x4a5040, {{0x39f01, 0x7}, 0x1})
	github.com/withastro/compiler/internal/printer/print-to-json.go:267 +0x4b
github.com/withastro/compiler/internal/printer.PrintToJSON({0x42a2a0, 0x66}, 0x4a5040, {{0x39f01, 0x7}, 0x1})
	github.com/withastro/compiler/internal/printer/print-to-json.go:114 +0xc
main.Parse.func1({{}, 0x7ff800010000000d, 0x410ea8}, {0x42c120, 0x2, 0x2})
	./astro-wasm.go:232 +0x2b
syscall/js.handleEvent()
	syscall/js/func.go:94 +0x26
exit code: 2
```

### Steps to Reproduce

1. The code is copied from Chinese version documentation: [https://docs.astro.build/zh-cn/core-concepts/astro-components/#%E7%BB%84%E4%BB%B6%E5%8F%82%E6%95%B0](https://docs.astro.build/zh-cn/core-concepts/astro-components/#%E7%BB%84%E4%BB%B6%E5%8F%82%E6%95%B0)

--- a/.changeset/kind-cheetahs-listen.md
+++ b/.changeset/kind-cheetahs-listen.md
@@ -1,0 +1,5 @@
+---
+'@astrojs/compiler': patch
+---
+
+Fixes a sourcemap-related crash when using multibyte characters

--- a/internal/sourcemap/sourcemap.go
+++ b/internal/sourcemap/sourcemap.go
@@ -616,7 +616,7 @@ func (b *ChunkBuilder) GetLineAndColumnForLocation(location loc.Loc) []int {
 	for count > 0 {
 		step := count / 2
 		i := originalLine + step
-		if lineOffsetTables[i].byteOffsetToStartOfLine <= location.Start {
+		if len(lineOffsetTables) > i && lineOffsetTables[i].byteOffsetToStartOfLine <= location.Start {
 			originalLine = i + 1
 			count = count - step - 1
 		} else {
@@ -629,7 +629,10 @@ func (b *ChunkBuilder) GetLineAndColumnForLocation(location loc.Loc) []int {
 	line := &lineOffsetTables[originalLine]
 	originalColumn := int(location.Start - line.byteOffsetToStartOfLine)
 	if line.columnsForNonASCII != nil && originalColumn >= int(line.byteOffsetToFirstNonASCII) {
-		originalColumn = int(line.columnsForNonASCII[originalColumn-int(line.byteOffsetToFirstNonASCII)])
+		newColumn := originalColumn - int(line.byteOffsetToFirstNonASCII)
+		if len(line.columnsForNonASCII) > newColumn {
+			originalColumn = int(line.columnsForNonASCII[newColumn])
+		}
 	}
 
 	// 1-based line, 1-based column

--- a/packages/compiler/test/parse/multibyte-characters.ts
+++ b/packages/compiler/test/parse/multibyte-characters.ts
@@ -1,0 +1,21 @@
+import { test } from 'uvu';
+import * as assert from 'uvu/assert';
+import { parse } from '@astrojs/compiler';
+
+const FIXTURE = `{foo}，`;
+
+test('does not crash', async () => {
+  const result = await parse(FIXTURE);
+  assert.ok(result.ast, 'does not crash');
+});
+
+test('properly maps the position', async () => {
+  const {
+    ast: { children },
+  } = await parse(FIXTURE);
+  const text = children[1];
+  assert.equal(text.position.start.offset, 5, 'properly maps the text start position');
+  assert.equal(text.position.end.offset, 8, 'properly maps the text end position');
+});
+
+test.run();

